stellargraph/layer/knowledge_graph.py

```diff
@@ -137,6 +137,96 @@ def embeddings(model):
 
         return node, rel
 
+    @staticmethod
+    def rank_edges_against_all_nodes(model, test_data, known_edges_graph):
+        """
+        Returns the ranks of the true edges in ``test_data``, when scored against all other similar
+        edges.
+
+        For each input edge ``E = (s, r, o)``, the score of the *modified-object* edge ``(s, r, n)``
+        is computed for every node ``n`` in the graph, and similarly the score of the
+        *modified-subject* edge ``(n, r, o)``.
+
+        This computes "raw" ranks: the score of each edge is ranked against all of the
+        modified-object and modified subject-ones, for instance, if ``E = ("a", "X", "b")`` has
+        score 3.14, and only one modified-object edge has a higher score (e.g. ``("a", "X", "c")``),
+        then the raw modified-object rank for ``E`` will be 2; if all of the ``(n, "X", "b")`` edges
+        have score less than 3.14, then the raw modified-subject rank for ``E`` will be 1.
+
+        Args:
+            model (tensorflow.keras.Model): a Keras model created using a ``ComplEx`` instance.
+
+            test_data: the output of :meth:`KGTripleGenerator.flow` on some test triples
+
+            known_edges_graph (StellarGraph):
+                a graph instance containing all known edges/triples
+
+        Returns:
+            A numpy array of integer raw ranks. It has shape ``N × 2``, where N is the number of
+            test triples in ``test_data``; the first column (``array[:, 0]``) holds the
+            modified-object ranks, and the second (``array[:, 1]``) holds the modified-subject
+            ranks.
+        """
+
+        if not isinstance(test_data, KGTripleSequence):
+            raise TypeError(
+                "test_data: expected KGTripleSequence; found {type(test_data).__name__}"
+            )
+
+        num_nodes = known_edges_graph.number_of_nodes()
+
+        def ranks(pred, true_ilocs, true_is_source):
+            batch_size = len(true_ilocs)
+            assert pred.shape == (num_nodes, batch_size)
+
+            # the score of the true edge, for each edge in the batch (this indexes in lock-step,
+            # i.e. [pred[true_ilocs[0], range(batch_size)[0]], ...])
+            true_scores = pred[true_ilocs, range(batch_size)]
+
+            # for each column, compare all the scores against the score of the true edge
+            greater = pred > true_scores
+
+            # the raw rank is the number of elements scored higher than the true edge
+            raw_rank = 1 + greater.sum(axis=0)
+            assert raw_rank.shape == (batch_size,)
+
+            return raw_rank
+
+        n, all_r = ComplEx.embeddings(model)
```

Comment:
just to make the connection between `n` and `all_r` clear below. Could you change this to:

 ```suggestion
        all_n, all_r = ComplEx.embeddings(model)
```

or maybe:

 ```suggestion
        all_node_embs, all_rel_embs = ComplEx.embeddings(model)
```

Reply:
Yeah, no need to be so brief. I'll go with the second 👍